### PR TITLE
Cache `canonical_matrix(::MonomialOrdering)`

### PR DESCRIPTION
**Pull request by lgoettgens**
Implements the proposal of https://github.com/oscar-system/Oscar.jl/issues/2498#issuecomment-1612689919.

The speed change is shown in https://github.com/oscar-system/Oscar.jl/issues/2498#issuecomment-1612720461.

Feel free to close this and propose an alternative to this @HechtiDerLachs.

### CODE REVIEW

File: src/Rings/orderings.jl [hunk 1]
```diff
@@ -169,11 +169,16 @@ end
 """
 Orderings actually applied to polynomial rings (as opposed to variable indices)
 """
-mutable struct MonomialOrdering{S}
+@attributes mutable struct MonomialOrdering{S}
   R::S
   o::AbsGenOrdering
   is_total::Bool
   is_total_is_known::Bool
```
Comment (HechtiDerLachs) [inline src/Rings/orderings.jl:176]:
```suggestion
  is_total_is_known::Bool
  canonical_matrix::ZZMatrix
```

File: src/Rings/orderings.jl [hunk 2]
```diff
@@ -1312,7 +1317,7 @@ julia> canonical_matrix(o2)
 [0   -1   0]
 ```
 """
-function canonical_matrix(M::MonomialOrdering)
+@attr ZZMatrix function canonical_matrix(M::MonomialOrdering)
   return canonical_matrix(nvars(base_ring(M)), M.o)
 end
```
Comment (HechtiDerLachs) [inline src/Rings/orderings.jl]:
```suggestion
function canonical_matrix(M::MonomialOrdering)
  if !isdefined(M, :canonical_matrix)
    M.canonical_matrix = canonical_matrix(nvars(base_ring(M)), M.o)
  end
  return M.canonical_matrix
end
```

File: src/Rings/orderings.jl [hunk 3]
```diff
@@ -169,11 +169,16 @@ end
 """
 Orderings actually applied to polynomial rings (as opposed to variable indices)
 """
-mutable struct MonomialOrdering{S}
+@attributes mutable struct MonomialOrdering{S}
```
Comment (HechtiDerLachs) [inline src/Rings/orderings.jl]:
```suggestion
mutable struct MonomialOrdering{S}
```

**Review by HechtiDerLachs — COMMENTED** [review]
I'd suggest not to go via attributes. This just entails another dictionary lookup and this is data structure is very low-level, i.e. it is called with high frequency.

**Comment by HechtiDerLachs** [comment]
And as @fieker said: Avoiding the lookup of a groebner basis in the first place is also possible and should be quicker.

But thanks for the very quick and constructive response!

File: src/Rings/orderings.jl (hunk 2 above)
Comment (fieker) [reply src/Rings/orderings.jl]: In general, if s.th. should be stored in almost all instances, I'd not put it in attributes, but into the structure. This allows for discovery...

**Comment by fingolfin** [comment]
This just hides over the actual problem, which is that `canonical_matrix` is being called far too often. See the discussion in #2099

**Comment by fingolfin** [comment]
Lest I be misunderstood: I still think this PR is useful and I am fine with merging it, but I removed the `Resolves #2498` because I think it would be a bad idea to keep fixing these "leave issues" instead of attacking the root.

File: src/Rings/orderings.jl [hunk 4]
```diff
@@ -174,6 +174,7 @@ mutable struct MonomialOrdering{S}
   o::AbsGenOrdering
   is_total::Bool
   is_total_is_known::Bool
+  canonical_matrix::ZZMatrix
 end
 
 function MonomialOrdering(R::S, o::AbsGenOrdering) where S
```
Comment (fingolfin) [inline src/Rings/orderings.jl]: I think you need to adjust this constructor to fix the errors: either make it an non-parametric inner constructor (so use `new{S}` in the next line); or else add a new inner parametric constructor that allows omitting the final field.

File: src/Rings/orderings.jl [hunk 5]
```diff
@@ -174,6 +174,7 @@ mutable struct MonomialOrdering{S}
   o::AbsGenOrdering
   is_total::Bool
   is_total_is_known::Bool
+  canonical_matrix::ZZMatrix
```
Comment (fingolfin) [inline src/Rings/orderings.jl:177]:
```suggestion
  canonical_matrix::ZZMatrix
  function MonomialOrdering(R::S, o::AbsGenOrdering) where S
     return new{S}(R, o, false, false)
  end
```
or for more flexibility:
```suggestion
  canonical_matrix::ZZMatrix
  function MonomialOrdering(R::S, o::AbsGenOrdering) where S
     return new{S}(R, o, false, false)
  end
  function MonomialOrdering(R::S, o::AbsGenOrdering, is_total::Bool) where S
     return new{S}(R, o, is_total, true)
  end
```
and then adjust the outer constructor below

**Comment by simonbrandhorst** [comment]
I am glad for anything that helps my computations to actually terminate.

**Review by simonbrandhorst — APPROVED** [review]

**Pull request merged**